l20n/migrate/transforms.py

```diff
@@ -0,0 +1,247 @@
+
+            placeable = FTL.Placeable(replacements[key])
+
+            # Return the elements found and converted so far, and the remaining
+            # text which hasn't been scanned for placeables yet.
+            return (
+                parts + [FTL.TextElement(before), placeable],
+                FTL.TextElement(after)
+            )
+
+        def is_non_empty(elem):
+            """Used for filtering empty `FTL.TextElement` nodes out."""
+            return not isinstance(elem, FTL.TextElement) or len(elem.value)
+
+        # Start with an empty list of elements and the original translation.
+        init = ([], FTL.TextElement(self.source))
+        parts, tail = reduce(replace, keys_in_order, init)
+
+        # Explicitly concat the trailing part to get the full list of elements
+        # and filter out the empty ones.
+        elements = filter(is_non_empty, parts + [tail])
+        quoted = self.source.startswith(' ') or self.source.endswith(' ')
+
+        return FTL.Pattern(None, elements, quoted)
+
+
+class PLURALS(Transform):
+    """Convert semicolon-separated variants into a select expression.
+
+    Build an `FTL.SelectExpression` with the supplied `selector` and variants
+    extracted from the `source`.  Each variant will be run through the
+    `foreach` function.  It should return an `FTL.Pattern`.
+    """
+
+    def __init__(self, source, selector, foreach):
+        self.source = source
+        self.selector = selector
+        self.foreach = foreach
+
+    def __call__(self, ctx):
+        variants = self.source.split(';')
+        keys_enum = enumerate(ctx.plural_categories)
+        last_variant_index = len(variants) - 1
+
+        def createMember(variant):
+            index, key = next(keys_enum)
+            # Run the legacy variant through `foreach` which returns an
+            # `FTL.Node` describing the transformation required for each
+            # variant.  Then evaluate it to a migrated FTL node.
+            value = evaluate(ctx, self.foreach(variant))
+            return FTL.Member(
+                FTL.Keyword(key),
+                value,
+                default=index == last_variant_index
+            )
+
+        select = FTL.SelectExpression(
+            self.selector,
+            variants=map(createMember, variants)
+        )
+        elements = [FTL.Placeable([select])]
+
+        return FTL.Pattern(None, elements, quoted=False)
+
+
+class CONCAT(Transform):
+    """Concatenate elements of many patterns."""
+
+    def __init__(self, *patterns):
+        self.patterns = list(patterns)
+
+    def __call__(self, ctx):
+        # Flatten the list of patterns of which each has a list of elements.
+        elements = [
+            elems for pattern in self.patterns for elems in pattern.elements
+        ]
+
+        # Merge adjecent `FTL.TextElement` nodes.
+        def merge_adjecent_text(acc, cur):
+            if len(acc) == 0 or type(cur) != FTL.TextElement:
```

Comment:
I totally read the logic wrong here. Is

    if not(len(acc) and type(cur) == FTL.TextElement):

Maybe the early return isn't as useful, and the code is actually easier to read if you put the main code into the if?